internal/querycoordv2/balance/score_based_balancer.go

```diff
@@ -0,0 +1,387 @@
+		plans := b.AssignChannel(dmChannels, onlineNodes)
+		for i := range plans {
+			plans[i].From = nodeID
+			plans[i].ReplicaID = replica.ID
+			plans[i].Weight = GetWeight(1)
+		}
+		channelPlans = append(channelPlans, plans...)
+	}
+	return channelPlans
+}
+
+func (b *ScoreBasedBalancer) getNormalSegmentPlan(replica *meta.Replica, nodesSegments map[int64][]*meta.Segment) []SegmentAssignPlan {
+	if b.scheduler.GetSegmentTaskNum() != 0 {
+		// scheduler is handling segment task, skip
+		return nil
+	}
+	segmentPlans := make([]SegmentAssignPlan, 0)
+	// generate candidates
+	nodeItems := b.convertToNodeItems(replica.GetCollectionID(), lo.Keys(nodesSegments))
+
+	minQueue := newPriorityQueue()
+	for _, item := range nodeItems {
+		minQueue.push(item)
+	}
+
+	maxQueue := newPriorityQueue()
+	for _, item := range nodeItems {
+		nodeItem := newNodeItem(-item.priority, item.nodeID)
+		maxQueue.push(&nodeItem)
+	}
+
+	for {
+		toNode := minQueue.pop()
+		fromNode := maxQueue.pop()
+
+		// pick the largest segment from fromNode, try to assign to toNode see if the cluster becomes more balance.
+		fromSegments := b.dist.SegmentDistManager.GetByCollectionAndNode(replica.CollectionID, fromNode.(*nodeItem).nodeID)
+		// sort the segments in asc order, try to mitigate to-from-unbalance
+		sort.Slice(fromSegments, func(i, j int) bool {
+			return fromSegments[i].GetNumOfRows() < fromSegments[j].GetNumOfRows()
+		})
+
+		// TODO we shouldn't use calculatePriority, because it it's balanced by replica, then global segment count will not be stable
+		// Better way is to calculate priority with a segment distribution map
+		// should not use the same snapshot from distribution throughout the process
+		fromPriority := -fromNode.(*nodeItem).priority
+		toPriority := toNode.(*nodeItem).priority
+
+		unbalance := fromPriority - toPriority
+		havingBalanced := false
+
+		updatePriority := func(nextToPriority int, toNode item, nextFromPriority int, fromNode item) {
+			havingBalanced = true
+			toNode.setPriority(nextToPriority)
+			minQueue.push(toNode)
+			fromNode.setPriority(-nextFromPriority)
+			maxQueue.push(fromNode)
+			fromPriority = nextFromPriority
+			toPriority = nextToPriority
+		}
+
+		for _, s := range fromSegments {
+			nextFromPriority := fromPriority - int(s.GetNumOfRows()) - int(float64(s.GetNumOfRows())*
+				params.Params.QueryCoordCfg.GlobalRowCountFactor.GetAsFloat())
+			nextToPriority := toPriority + int(s.GetNumOfRows()) + int(float64(s.GetNumOfRows())*
+				params.Params.QueryCoordCfg.GlobalRowCountFactor.GetAsFloat())
+			if nextToPriority <= nextFromPriority {
+				plan := SegmentAssignPlan{
+					ReplicaID: replica.GetID(),
+					From:      fromNode.(*nodeItem).nodeID,
+					To:        toNode.(*nodeItem).nodeID,
+					Segment:   s,
+					Weight:    GetWeight(0),
+				}
+				segmentPlans = append(segmentPlans, plan)
+				updatePriority(nextToPriority, toNode, nextFromPriority, fromNode)
```

Comment:
This method maybe called many time in this loop and will push nodes to queue many times.

Reply:
BTW, I think this balance algorithm can not reach a balanced state. Think about a simple case, that nodes have priority 1 , 2 and 3 and every segments's priority is 1. Then the minQueue's init state is 1,2,3 and maxQueue is -3, -2, -1. After the first balance round, minQueue will be 2,2,3 and maxQueue will be -2. -2, -1. And the final state will be 2,3,3 and -2,-1,-1(because a single node is not updated in two queues at the same time). And the nodes priority are 3,2,1 in the end, but we want it to be 2,2,2.

Reply:
right, we should probably has another implementation for the queue
maybe just sort the queue for each time would work